### ISSUE TITLE
Feature request - rule for await in async try catch return

### ISSUE THREAD

**Issue by drabinowitz** [typescript-eslint/typescript-eslint#994]
A `Rejected Promise` returned by an `async` function will not be caught by the function's `try-catch` block. This may surprise some users and lead to unexpected behavior:
```typescript
(async () => {
  try {
    // will result in an UnhandledPromiseRejection
    return Promise.reject('Error!');
  } catch (e) {
    // will not run
    handleCleanup();
  }
})();
```

conversely, `await`-ed promises work in a more intuitive fashion:
```typescript
(async () => {
  try {
    // will not result in an UnhandledPromiseRejection
    return await Promise.reject('Error!');
  } catch (e) {
    // will run
    handleCleanup();
  }
})();
```

I understand why this is the case according to the `async-await` function spec, but it feels like confusing behavior and a problem that `typescript-eslint` could help avoid via the following rule:

disallow `return`-ing a `Promise` from inside an `async` function `try-catch` block.

Thus, the first case described above would be invalid according to this rule, but the second case would be valid. We could even, potentially, provide an auto-fix that would add the `await` statement to the promise.

There's also the possibility for the broader rule of disallowing`return`-ing a `Promise` from anywhere inside an `async` function, but I wanted to start with the narrowest case.

Please let me know if there's any other clarifying information I can provide, thanks so much for your time!

**Comment by drabinowitz** [comment]
Yes i think that'd make more sense @ark120202. @bradzacher what are you thoughts on that change should I update the feature request?

**Comment by bradzacher** [comment]
To clarify then,

The idea would be to have a rule (`return-await`) which bans directly returning await statements (`return await foo`) (i.e. requires no type info), except in the case that you're returning a promise from within a try/catch, in which case you want to enforce that you return an await (i.e. requires type info).

Seems like a good candidate for this plugin.

```
declare function returnsPromise(): Promise<string>;
declare function handleException(e: Error): string;

// option: 'in-try-catch'
async function myFunc1(): Promise<string> {
  try {
    return returnsPromise(); // lint error - should await the promise so errors are handled by the surrounding try/catch
  } catch (e) {}
  return Promise.resolve('foo'); // no error
}

// option: 'never'
async function myFunc1(): Promise<string> {
  try {
    return await returnsPromise(); // lint error - should never return await
  } catch (e) {}
  return await Promise.resolve('foo'); // lint error 
}

// option: 'always'
async function myFunc1(): Promise<string> {
  try {
    return returnsPromise(); // lint error - should always await promise
  } catch(e) {}
  return Promise.resolve('foo'); // lint error 
}
```

**Comment by drabinowitz** [comment]
that's what i was thinking, yeah! @ark120202 is the snippet above what you were thinking for this?

**Comment by ark120202** [comment]
Yeah, that sounds good to me!
One thing I'm not really sure is `'never'`, since currently `no-return-await` doesn't disallow it in try catch (but doesn't enforce it as well). So if compatibility with `no-return-await` is one of goals it has to be allowed.

**Comment by drabinowitz** [comment]
I'm actually okay with that since the same could be said for the `always` rule (in terms of incompatibility with `no-return-await`). I think if we make the default `in-try-catch` then we avoid this issue for most users. It feels appropriate to expect users with customized rules to have to be more careful in understanding how those rules will interact.

**Comment by drabinowitz** [comment]
@bradzacher so in terms of next steps should i put together a pr that references this issue? I haven't worked on a codebase like this in the past but I'd be happy to give it a go

**Comment by ark120202** [comment]
Actually, in that case is there a need for `'never'` at all? I have originally proposed it to be more compatible with `no-return-await`, but I agree that there's no reason to pursue that. Would someone really want to disallow `return await` even in try catch?

**Comment by drabinowitz** [comment]
hmm interesting thought. Performance critical applications i suppose, but even then this is a pretty narrow optimization. @bradzacher did you have a particular use case in mind or just for consistency purposes?

**Comment by bradzacher** [comment]
consistency, warding off people asking for it.

I can pretty easily see an argument that people don't want return await ever, and would rather a variable assignment that gets returned a la
```ts
async function myFunc1(): Promise<string> {
  try {
    const res = await returnsPromise();
    return res;
  } catch (e) {}
  const res = await Promise.resolve('foo');
  return res;
}
```

Not sure - it's probably the easiest case to support, so I figure there's little overhead to supporting it.

**Comment by drabinowitz** [comment]
so in terms of next steps for this, should I open a PR?

**Comment by bradzacher** [comment]
Definitely - if you see value in this and want it sooner, rather than later!

It's always a lot faster and better for us if users PR the features we want; either we maintainers can try to write PRs for the ~180 issues in the repo, or we can review PRs for those issues. The latter is more scalable and easier.

**Comment by drabinowitz** [comment]
sounds good ill get to work on this, this weekend!
